internal/olm/operator/registry/index_image.go

```diff
@@ -121,14 +157,279 @@ func (c IndexImageCatalogCreator) CreateCatalog(ctx context.Context, name string
 	return cs, nil
 }
 
+// getChannelHead retrieves the channel head from an array of entries
+func getChannelHead(entries []declarativeconfig.ChannelEntry) (string, error) {
+	nameMap := make(map[string]bool)
+	replacesMap := make(map[string]bool)
+
+	for i := range entries {
+		nameMap[entries[i].Name] = true
+		if entries[i].Replaces != "" {
+			replacesMap[entries[i].Replaces] = true
+		}
+	}
+
+	// gets the CSV name that does not appear in any replaces field in the entries array
+	for key := range nameMap {
+		if _, present := replacesMap[key]; !present {
+			return key, nil
+		}
+	}
+
+	// this should not be reached since there must be an edge to upgrade from
+	return "", errors.New("no channel head found")
+}
+
+// handleTraditionalUpgrade upgrades an operator that was installed using OLM. Subsequent upgrades will go through the runFBCUpgrade function
+func handleTraditionalUpgrade(ctx context.Context, indexImage string, bundleImage string, channelName string) (string, error) {
+	// render the index image
+	originalDeclCfg, err := renderRefs(ctx, []string{indexImage})
+	if err != nil {
+		return "", err
+	}
+
+	// render the bundle image
+	bundleDeclConfig, err := renderRefs(ctx, []string{bundleImage})
+	if err != nil {
+		return "", err
+	}
+
+	if len(bundleDeclConfig.Bundles) != 1 {
+		return "", errors.New("bundle image must have at least one bundle")
+	}
+
+	// search for the specific channel in which the upgrade needs to take place, and upgrade from the channel head
+	for i := range originalDeclCfg.Channels {
+		if originalDeclCfg.Channels[i].Name == channelName && originalDeclCfg.Channels[i].Package == bundleDeclConfig.Bundles[0].Package {
+			// found specific channel
+			channelHead, err := getChannelHead(originalDeclCfg.Channels[i].Entries)
+			if err != nil {
+				return "", err
+			}
+			entry := declarativeconfig.ChannelEntry{
+				Name:     bundleDeclConfig.Bundles[0].Name,
+				Replaces: channelHead,
+			}
+			originalDeclCfg.Channels[i].Entries = append(originalDeclCfg.Channels[i].Entries, entry)
+			break
+		}
+	}
+
+	// add the upgraded bundle to resulting declarative config
+	originalDeclCfg.Bundles = append(originalDeclCfg.Bundles, bundleDeclConfig.Bundles[0])
+
+	// validate the declarative config and convert it to a string
+	var content string
+	if content, err = ValidateAndStringify(originalDeclCfg); err != nil {
+		return "", fmt.Errorf("error validating and converting the declarative config object to a string format: %v", err)
+	}
+
+	log.Infof("Generated a valid Upgraded File-Based Catalog")
+
+	return content, nil
+}
+
+// runFBCUpgrade starts the process of upgrading a bundle in an FBC. This function will recreate the FBC that was generated
+// during run bundle and upgrade a specific bundle in the specified channel.
+func runFBCUpgrade(ctx context.Context, c *IndexImageCatalogCreator) error {
```

Comment:
Seems better semantically and it's more common practice in Go to define this as a struct receiver as opposed to passing it as a parameter. 

```suggestion
func (c IndexImageCatalogCreator) runFBCUpgrade(ctx context.Context) error {
```